### DISCUSSION TITLE
V2 General Discussion

### DISCUSSION THREAD

**Discussion by taniabogatsch** [marcboeker/go-duckdb#232] [General]
Currently, go-duckdb tries to avoid any breaking changes. However, in a recent discussion about opt-in vs. opt-out [here](https://github.com/marcboeker/go-duckdb/pull/205) for additional driver features, we concluded that some breaking changes when moving to a new version are beneficial.

I am opening this discussion to address possible changes and improvements relevant for a V2. 

Related discussion addressing static library builds: https://github.com/marcboeker/go-duckdb/discussions/349

**Comment by marcboeker** [comment]
For v2 we should also think about the pre-built libraries as they are getting larger and larger over time. On the other hand, building DuckDB on the fly takes some time and a proper build chain. An fetching pre-built libraries from a different source is not Go-like.

**Comment by dxh9845** [comment]
I have been using this library for work. In general, I think the library works decently well, but I do think the following improvements would help ergonomics:

* Refactor Apache Arrow dependency to be opt in. It could possibly be a separate go module, or enabled explicitly with build tags. I think Golang Migrate does a good model of how this could work possibly?
* Improve the documentation around Appender. Currently the appender is pretty confusing to use - the docs mention you should use .Close, but it seems like .Flush is also required to make things work? I could be wrong.
* More examples within the Go Doc. Right now the examples are spare and usage can really only be figured out via using the library and seeing what works / what doesn't.

Pipe dream, but probably not possible:
* Remove the C bindings and make this library native Go. I don't think this is realistic, but ideally users could use this library without CGO and could build statically compiled binaries.

**Reply by james-lawrence** [reply]
I suspect removing cgo is possible via wasi but its a very large lift; but it would resolve the multiple platform binary build problem. there is a sqlite3 library that did this.

**Comment by mlafeldt** [comment]
👋 Thanks for starting this discussion!

This repository has moved to https://github.com/duckdb/duckdb-go

If this topic is still relevant, please continue the conversation at the new location. I'm closing this now as part of the migration.

Please see https://github.com/marcboeker/go-duckdb/issues/565 for more details.
